### PR TITLE
DRILL-6230: Extend row set readers to handle hyper vectors

### PR DESCRIPTION
**Pull request by paul-rogers**
The current row set readers have incomplete support for hyper-vectors. To add full support, we need an interface that supports either single batches or hyper batches. Accessing vectors in hyper batches differs depending on whether the vector is at the top level or is nested. See this post for details. Also includes a simpler reader template: replaces the original three classes with one, in parallel with the writers.

Key changes:

* Refactor the readers to generate just the required reader, then build up the optional and repeated readers as layers on top of the generated reader. This is the same structure that the writers already use.
* Add and test support for hyper-vectors.
* Extend the existing "vector accessor" abstraction to fully support the highly complex process of locating nested vectors (those within a map or union) in a hyper-batch.
* Introduce the idea of a "null state" abstraction to handle the messy null handling in unions and repeated lists.
* Modifies tests as needed for the new internal format of vector readers.

To keep the PR from getting overly large, this PR strips out the actual union and list support. That support will be added in a future PR. Similarly, there are matching changes to writers that will also be done in a separate PR.

Other minor changes:

* Revises the previous utility PR. In some cases, it turns out to be cleaner to use a separate `mapValue()` function instead of `objArray()`, even though both produce an object array. Calling it `mapValue()` makes it a bit clearer what we're trying to accomplish.

This PR is not needed for Drill 1.13; it can go into Drill 1.14.

See [this post](https://github.com/paul-rogers/drill/wiki/Batch-Handling-Upgrades) for details of the end-state toward which this PR is one step.

### CODE REVIEW

**Comment by paul-rogers** [comment]
@ppadma, I believe you are working in this area, can you please take a look at your convenience? Thanks.

**Comment by ppadma** [comment]
@paul-rogers I started the review. will get back soon.

**Comment by paul-rogers** [comment]
@ppadma , any partial comments that I can start to take a look at? There are still a number of PRs in this chain and it would be great if we could keep things ticking along... Thanks!

File: exec/java-exec/src/test/java/org/apache/drill/test/rowSet/HyperRowSetImpl.java [hunk 1]
```diff
@@ -45,8 +50,67 @@ public RowSetReader buildReader(HyperRowSet rowSet, SelectionVector4 sv4) {
...
+    @Override
+    public HyperRowSet build() throws SchemaChangeException {
+      SelectionVector4 sv4 = new SelectionVector4(allocator, totalRowCount);
+      ExpandableHyperContainer hyperContainer = new ExpandableHyperContainer();
+      for (VectorContainer container : batches) {
+        hyperContainer.addBatch(container);
+      }
+
+      // TODO: This has a bug. If the hyperset has two batches with unions,
+      // and the first union contains only VARCHAR, while the second contains
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/test/rowSet/HyperRowSetImpl.java]: [DRILL-6304](https://issues.apache.org/jira/browse/DRILL-6304)

File: exec/java-exec/src/test/java/org/apache/drill/test/rowSet/test/TestHyperVectorReaders.java [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+/**
+ * Test the reader mechanism that reads rows indexed via an SV4.
+ * SV4's introduce an additional level of indexing: each row may
+ * come from a different batch. The readers use the SV4 to find
+ * the root batch and vector, then must navigate downward from that
+ * vector for maps, repeated maps, lists, unions, repeated lists,
+ * nullable vectors and variable-length vectors.
+ * <p>
+ * This test does not cover repeated vectors; those tests should be added.
```
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/test/java/org/apache/drill/test/rowSet/test/TestHyperVectorReaders.java]: [DRILL-6305](https://issues.apache.org/jira/browse/DRILL-6305)

**Comment by paul-rogers** [comment]
Rebased on latest master. Ran Maven unit tests; they passed up to Mongo.

@ppadma, please run against the pre-commit tests

@parthchandra, can you give this one a committer review? It builds on some commits you looked at previously.

Thanks!

**Comment by ppadma** [comment]
@paul-rogers ran the tests. they all pass.

**Pull request closed without merging**